runtime/jvmti/jvmtiHelpers.c

```diff
@@ -1660,3 +1718,143 @@ ensureHeapWalkable(J9VMThread *currentThread)
 		}
 	}
 }
+
+#if JAVA_SPEC_VERSION >= 19
+static void
+tlsNullFinalizer(void *entry)
+{
+	/* do nothing */
+}
+#endif /* JAVA_SPEC_VERSION >= 19 */
+
+/**
+ * @brief Allocate a thread local storage key (which represents a jvmtiEnv) to index into a thread
+ * object's TLS array.
+ * @param[in] vm java vm
+ * @param[out] handle on successful return contains the allocated key value
+ * @return 0 on success, -1 on failure
+ */
+IDATA
+jvmtiTLSAlloc(J9JavaVM *vm, UDATA *handle)
+{
+#if JAVA_SPEC_VERSION >= 19
+	return jvmtiTLSAllocWithFinalizer(vm, handle, tlsNullFinalizer);
+#else /* JAVA_SPEC_VERSION >= 19 */
+	return omrthread_tls_alloc(handle);
+#endif /* JAVA_SPEC_VERSION >= 19 */
+}
+
+#if JAVA_SPEC_VERSION >= 19
+/**
+ * @brief Allocate a thread local storage key (which represents a jvmtiEnv) to index into a thread
+ * object's TLS array.
+ * @param[in] vm java vm
+ * @param[out] handle on successful return contains the allocated key value
+ * @param[in] finalizer a finalizer function which will be invoked when a thread is detached or terminates
+ * 		if the thread's TLS entry for this key is non-NULL
+ * @return 0 on success, -1 on failure
+ */
+IDATA
+jvmtiTLSAllocWithFinalizer(J9JavaVM *vm, UDATA *handle, j9_tls_finalizer_t finalizer)
+{
+	IDATA i;
+
+	*handle = 0;
+
+	omrthread_monitor_enter(vm->tlsFinalizersMutex);
+	for (i = 0; i < J9JVMTI_MAX_TLS_KEYS; i++) {
+		if (NULL == vm->tlsFinalizers[i]) {
+			*handle = i + 1;
+			vm->tlsFinalizers[i] = finalizer;
+			break;
+		}
+	}
+	omrthread_monitor_exit(vm->tlsFinalizersMutex);
+
+	return i < J9JVMTI_MAX_TLS_KEYS ? 0 : -1;
+}
+#endif /* JAVA_SPEC_VERSION >= 19 */
+
+/**
+ * @brief Free a thread local storage key allocated with jvmtiTLSAlloc.
+ * @param[in] vm java vm
+ * @param[in] key the key to be freed
+ * @return 0 on success, -1 on failure
+ */
+IDATA
+jvmtiTLSFree(J9JavaVM *vm, UDATA key)
+{
+#if JAVA_SPEC_VERSION >= 19
+	pool_state state;
+	J9JVMTIThreadData **each = NULL;
+
+	omrthread_monitor_enter(vm->tlsPoolMutex);
+	each = pool_startDo(vm->tlsPool, &state);
+	while (NULL != each) {
+		/* The pointer at this index has already been freed by pool_kill. */
```

Comment:
I don't understand the comment - probably best to just remove it - it's not the job of the TLS manager to worry about the contents of the slots.